servo/connectors/olas/controller.py

```diff
@@ -0,0 +1,880 @@
+                 )
+        servo.logger.info(stats)
+        if self.cfg.config.devel:
+            self.logfd.write(f"{self.ts-self.base_ts:.0f} {self.cpu:.1%} {self.pod_rate/self.cpu if self.cpu else math.nan:.1f} {self.total_rate:.1f} {self.pod_rate:.1f} {self.total_predict/self.replicas if self.total_predict else 0:.1f}"
+                             f" {self.target:.1f} {self.p50:.1f} {self.rq_time:.1f} {numpods} {self.replicas}\n")
+            self.logfd.flush()
+
+        fn = self.mode_table[self.mode]
+        if asyncio.iscoroutinefunction(fn):
+            await fn(metrics)
+        else:
+            fn(metrics)
+
+        if self.cfg.objectives.maxCost:
+            self.pod_cost = self.caculate_pod_cost(metrics)
+        servo.logger.info(f"scaler action {self.scale_request}")
+        if self.scale_request:
+            self.scale_request.sort()
+            replicas, raw, reason = self.scale_request[-1]
+            await self.output_scale_with_constrain(replicas, raw)
+        self.last_control_ts = self.ts
+
+    async def process_event_task(self):
+        queue = self.eventqueue
+        while 1:
+            event_type, metrics = await queue.get()
+            servo.logger.info("Process event {metrics}")
+            await self.control_with_metrics(metrics)
+
+    async def control_task(self):
+
+        self.boot()
+        self.eventqueue = eventqueue.eventQueue.subscribe_event("OLASMetrics")
+        asyncio.create_task(self.process_event_task())
+
+        if self.upload_cfg:
+            await self.cached_upload("config", self.cfg.dict())
+
+        if self.cfg.config.devel:
+            text = "ts cpu cpucap total-qps pod-qps prediction target p50 p90 n replicas"
+            print(text, file=self.logfd)
+
+        pods = await self.get_pods()
+        await self.collect_metrics(pods)
+
+        while 1:
+            await self.sleep_loop_interval()
+            if self.cfg.config.devel:
+                self.try_reload()
+            pods = await self.get_pods()
+            metrics = await self.collect_metrics(pods)
+            eventqueue.eventQueue.put_event("OLASMetrics", metrics)
+
+    def try_reload(self):
+        import importlib
+        import os
+        import sys
+        mtime = os.path.getmtime("olas/controller.py")
+        if self.reload_mtime and mtime != self.reload_mtime:
+            servo.logger.info(f"=====Reloading module controller classid: {id(self.__class__)} =======")
+            module = sys.modules['controller']
+            ctl = importlib.reload(module)
+            self.__class__ = ctl.OLASController
+            self.mode_table[mode.CPU_PREDICT] = getattr(self, "scale_by_cpu_predict")
+            servo.logger.info(f"{module} new classid: {id(self.__class__)}")
+
+            importlib.reload(prometheus)
+            importlib.reload(envoy_stats)
+            importlib.reload(client)
+            importlib.reload(kube)
+        self.reload_mtime = mtime
```

Comment:
This is a bad implementation of code reloading. I would recommend [Watchgod](https://pypi.org/project/watchgod/) instead

Reply:
@theStrenuous I recommend disabling this until / unless ServoX has a mechanism for this